Players make moves, board validation, script automation

KrzysiekJodlowski opened the pull request:
- start implementing basic game logic with Game, Players and Settings
- implement board validation mechanism
- add tests checking board validation (these one need some refactor in terms of mocking)
- reorganise code into packages
- reimplement scripts and add python script running scenarios, writing its results into files and comparing that files with template ones

KrzysiekJodlowski committed:
haha

KrzysiekJodlowski committed:
haha2

KrzysiekJodlowski committed:
Merge branch 'dev' of https://github.com/KrzysiekJodlowski/OX into dev

KrzysiekJodlowski committed:
Implement NumberRangeTest

Tests for representative range vs number to check corner cases.

KrzysiekJodlowski committed:
Merge branch 'bugfix' into dev

KrzysiekJodlowski committed:
Merge branch dev into players_make_moves

KrzysiekJodlowski committed:
Implement players make moves

Without new tests and checkstyle refactor

Commit:
Change docs according to checkstyle requirements

Commit:
Correct generic type reference in NumberRange

Commit:
Fixing bugs

- chane exception type in ConsoleUI method
- reimplement makeMove in HumanPlayer
- implement equals in Move
- use proper generic types in NumberRange
- reimplement main loop logic in OxConsoleGame

Commit:
Start implementing horizontal win check

KrzysiekJodlowski committed:
Implement win condition checking in four directions

- appropriate tests are added
- due to code duplication it would be good to reuse it (functional interfaces?)
- also tests need some refactor (mocking board etc)

KrzysiekJodlowski committed:
Extend checkers condition

Symbol check is crutial when defining which player wins.

KrzysiekJodlowski committed:
Add simplest draw game check scenario.

Also containsField was reimplemented as it requires only FieldNumber to work.

KrzysiekJodlowski committed:
Setting which player shpuld start

- reimplement Builder in Players
- add fields in draw test

KrzysiekJodlowski committed:
Rebuild automatic tests

- refresh of bash scripts are made
- new folder automata/ with python script running default scenarios
- do some checkstyle refactoring

KrzysiekJodlowski committed:
Merge branch playground with players_make_moves

And some really minor refactor.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Move.java:
```diff
@@ -0,0 +1,45 @@
+    return this;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Due to errorProne:
- Classes that override equals should also override hashCode.

KrzysiekJodlowski replied:
Thats right, done

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxGameEventBus.java:
```diff
@@ -0,0 +1,32 @@
+package com.github.krzysiekjodlowski.ox;
+
+import com.github.krzysiekjodlowski.ox.board.BoardValidator;
```
Unused import.

KrzysiekJodlowski replied:
Removed.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
+    Scanner scanner = setScanner(args);
+    // in the future releases (when more UI implementations will appear) another
+    // helper method checking UI chosen type will initialize the interface
+    UI<String, Integer> ui = new ConsoleUI(scanner, new PrintStream(System.out));
```
Due to SpotBugs:
- Missing charset in print stream

KrzysiekJodlowski replied:
Added:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
+        System.out.println("File not found!");
+      }
+    } else {
+      scanner = new Scanner(System.in);
```
Due to SpotBugs:
- Missing charset in scanner

KrzysiekJodlowski replied:
Added:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+
+}
```
Do you really need "this" keyword all over the code?

KrzysiekJodlowski replied:
No i don't, i can remove them if you want:) That's just my habit - I like explicitly point out that I'm using an instance variable.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+        Settings.MIN_GAME_BOARD_SIDE_LENGTH, Settings.MAX_GAME_BOARD_SIDE_LENGTH));
+    Settings.gameBoardSideLength = this.ui.getNumberFromUser(
+        NumberRange.of(Settings.MIN_GAME_BOARD_SIDE_LENGTH, Settings.MAX_GAME_BOARD_SIDE_LENGTH));
+    return new Board(Settings.gameBoardSideLength);
```
Unnecessary assignment of the board side length from user. You can pass that value directly to Board constructor and you don't need to store it in settings. Also it breaks single responsibility principle.

KrzysiekJodlowski replied:
You're right, it's getting tricky in terms of counting board capacity and board encapsulation, but I'll fix that.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    return new Board(Settings.gameBoardSideLength);
+  }
+
+  private Integer setWinCondition() {
```
Method returns something, despite this it has "set" prefix!

hongshui replied:
True. Normally you should avoid mixing accessors and mutators. Also, be careful with all names with "set" and "get" - if you are not simpy assigning to field / reading from field, you could try to chose some other name

KrzysiekJodlowski replied:
Well spotted! It's more a getter, as with the board side getter method (although both will be renamed as get_FromUser)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    this.ui.say(board.toString());
+    this.ui.say("Choose one field by entering its number "
+        + "(whole number; or press ctrl+C to exit game):");
+  }
```
Method name suggests that it breaks single responsibility principle.
Rename it (maybe something like "showNextMove" or something) or consider additional methods.

hongshui replied:
I think renaming would be enough. I don't feel like method itself is breaking SRP

AleksanderStudnicki replied:
In fact method doesn't break it (you need these actions together to show next move in ui) but its name suggests something different. That was my point.

KrzysiekJodlowski replied:
Done:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    } else {
+      this.ui.say(String.format("The winner is %s!", boardValidator.getWinner()));
+    }
+  }
```
Same as previous comment. Single responsibility principle.

KrzysiekJodlowski replied:
Done:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    }
+  }
+
+
```
Empty lines.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/board/Board.java:
```diff
@@ -0,0 +1,104 @@
+      try {
+        currentFieldNumber = FieldNumber.valueOf(i);
+      } catch (NumberLowerThanOneException e) {
+        e.printStackTrace();
```
Not nice.

KrzysiekJodlowski replied:
I know, problem is that in this case, if nobody will change this class (ok it can be a weak explanation;) the iteration starts from one, always. I'll change it printing exception message.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Players.java:
```diff
@@ -0,0 +1,61 @@
+    private Player startingPlayer = this.playerOne;
+
+    public Builder() {
+    }
```
There are no other constructors so no need of this.

KrzysiekJodlowski replied:
True, removed.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Settings.java:
```diff
@@ -0,0 +1,26 @@
+  static final int MIN_GAME_BOARD_SIDE_LENGTH = 3;
+  static final int MAX_GAME_BOARD_SIDE_LENGTH = 40;
+  static final int FIRST_FIELD_NUMBER = 1;
+  static int gameBoardSideLength = 3;
```
There is no need of storing this value in Settings. I mentioned it in OXConsoleGame class comment.
It would be dangerous if we want to handle multiple games at once because Settings is singleton and each game would be modifying value of this variable.

KrzysiekJodlowski replied:
Ok, I've removed boardSideLength and boardCapacity() from settings, but in cost of providing a public getter for capacity. I could count boardSideLength * boardSideLength gained from user tough.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Settings.java:
```diff
@@ -0,0 +1,26 @@
+   */
+  public int boardCapacity() {
+    return gameBoardSideLength * gameBoardSideLength;
+  }
```
In my opinion unnecessary but HumanPlayer class uses this method. I think boardCapacity is a property of the Board class! In that case HumanPlayer class should somehow calls
 method/get variable from the Board class not from the Settings singleton. You should consider how to change that.

KrzysiekJodlowski replied:
Removed:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/board/SlantChecker.java:
```diff
@@ -0,0 +1,68 @@
+            FieldNumber.valueOf(nextField)).equals(playersMove.getPlayersSymbol()
+    );
+  }
+}
```
Due to CPD (part of PMD):
- Code duplication (same as ReversedSlantChecker).

KrzysiekJodlowski replied:
checkIfTherIsAnother moved to abstract class extended both by SlantChecker and ReverseSlantChecker:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/board/ReverseSlantChecker.java:
```diff
@@ -0,0 +1,67 @@
+            FieldNumber.valueOf(nextField)).equals(playersMove.getPlayersSymbol()
+    );
+  }
+}
```
Due to CPD (part of PMD):
- Code duplication (same as SlantChecker)

KrzysiekJodlowski replied:
Same as above

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Settings.java:
```diff
@@ -0,0 +1,26 @@
+  static final int MIN_GAME_BOARD_SIDE_LENGTH = 3;
+  static final int MAX_GAME_BOARD_SIDE_LENGTH = 40;
+  static final int FIRST_FIELD_NUMBER = 1;
+  static int gameBoardSideLength = 3;
```
No need of "static" keyword (enum properties!).

KrzysiekJodlowski replied:
True, removed:)

AleksanderStudnicki commented on pom.xml:
```diff
@@ -28,8 +28,8 @@
         <version.dependency.mockito>3.2.4</version.dependency.mockito>
         <!-- plugin versions should go here, so below they can be omitted -->
         <version.plugin.maven.enforcer>3.0.0-M2</version.plugin.maven.enforcer>
```
There is a newer version (M3).

KrzysiekJodlowski replied:
Missed that one! Updated.

AleksanderStudnicki commented on automata/run_auto.py:
```diff
@@ -0,0 +1,67 @@
+        file_out.write("\n\n== Total error count: {}\n\n".format(total_errors))
+
+run_scenarios()
+validate_scenarios()
```
Doesn't work for me.

Traceback (most recent call last):
  File "automata/run_auto.py", line 11, in <module>
    output_dir = create_output_dir()
  File "automata/run_auto.py", line 9, in create_output_dir
    os.mkdir(output)
FileNotFoundError: [Errno 2] No such file or directory: 'automata/output/2020-01-06_18-05/'

KrzysiekJodlowski replied:
Did you run the script itself or via auto.sh? You should choose second option (as it states in README file), in first case I have the same result.

AleksanderStudnicki replied:
I run script from auto.sh and it still doesn't work. FYI I'm using MacOS 10.14. I can check that tomorrow on Ubuntu 18.

KrzysiekJodlowski replied:
Maybe its connected with not properly handled exceptions, I'll check that tomorrow morning:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Player.java:
```diff
@@ -0,0 +1,61 @@
+   * @param ui generic user interface
+   * @return Move made by player
+   */
+  public abstract Move makeMove(UI<String, Integer> ui) throws NumberLowerThanOneException;
```
UI in player class. It doesn't look nice for me. I think you should consider delegate such a method to a game class. In my opinion Player should be just a class to represent player's data (symbol, name, score - that kind of stuff) with handling basic operations like adding win or draw. When you make a move you need some board properties (maximum index of board) and player doesn't hold these properties and also doesn't hold a reference to the board in your implementation. Your game class implementation (OxConsoleGame) has the board so I think you can do it with a little effort.

KrzysiekJodlowski replied:
You've got the point here, my bad. Moved collecting move logic to game class.

hongshui commented on automata/templates/O_wins_first_column.txt:
```diff
@@ -0,0 +1,39 @@
+*** Welcome to OX game! ***
+What size of a game board side do you choose? (whole numbers from 3 to 40)
```
This message is a bit confusing. Maybe something simpler like "Please select board's side length" or something?

KrzysiekJodlowski replied:
Your'e right - corrected.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Players.java:
```diff
@@ -0,0 +1,61 @@
+  public void handle(Event<?> event) {
+    this.currentPlayer = this.currentPlayer.equals(this.playerOne)
+        ? this.playerTwo : this.playerOne;
+  }
```
I found no usage of this method. Maybe you should not implement Subscriber for that class.

KrzysiekJodlowski replied:
There must be logic of changing player somewhere. As Players keep both players I found it natural change the current player there. But I will remove Subscriber implementation and provide nextPlayer method instead:)

hongshui commented on automata/templates/O_wins_first_column.txt:
```diff
@@ -0,0 +1,39 @@
+ 4 5 6
+ 7 8 9
+
+Choose one field by entering its number (whole number; or press ctrl+C to exit game):
```
Repeating "or press ctrl+C" with every move seems unnecassary. Maybe print this info once, at the very start, and do not repeat later? Also, maybe just "Please enter field number" or something like this? Your current message seems also a bit to long

hongshui commented on automata/templates/O_wins_first_column.txt:
```diff
@@ -0,0 +1,39 @@
+ O X 6
+ O 8 9
+
+Demo 'game' is over!
```
Do not put game in '', game is game ;)

KrzysiekJodlowski replied:
Corrected:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/board/Board.java:
```diff
@@ -0,0 +1,104 @@
+  private boolean containsMarkedField(final FieldNumber fieldNumber) {
+    return this.fields.containsKey(fieldNumber);
+  }
+}
```
"This" keyword everywhere.

KrzysiekJodlowski replied:
As I said, I can remove all this'es but I didn't recieved any reprmands from other coworkers yet;)

hongshui commented on automata/templates/O_wins_first_column.txt:
```diff
@@ -0,0 +1,39 @@
+*** Welcome to OX game! ***
+What size of a game board side do you choose? (whole numbers from 3 to 40)
+Which player should start? (enter 1 if player 'O' and 2 if player'X')
+Player O (human)
```
It was not a human ;) I'm not sure if printing who is moving (human or machine) is really neccessary

KrzysiekJodlowski replied:
It was a human, only he made his moves a bit earlier when creating a script;)

hongshui commented on readme.adoc:
```diff
@@ -13,8 +13,18 @@ To run project you have to use:
-- if you can't run a script first execute command like: *_chmod 755 custom.sh_*
-
+- project with automated input run _"auto"_ script
+- running script from command line by typing *_./standard.sh_* or *_./auto.sh_*
```
run not running

KrzysiekJodlowski replied:
Ah, my bad, corrected:)

hongshui commented on readme.adoc:
```diff
@@ -13,8 +13,18 @@ To run project you have to use:
-- running script from command line by typing eg: *_./custom.sh (custom_board_sizes.input)_*
-- if you can't run a script first execute command like: *_chmod 755 custom.sh_*
-
+- project with automated input run _"auto"_ script
```
Aren't this two messages repeating what is described in third line or have I understood something wrong?

KrzysiekJodlowski replied:
Not exactly, first two describes what two scripts does, third explains how to run it, and fourth describes what to do when user can't run a script.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/model/Symbol.java:
```diff
@@ -0,0 +1,26 @@
+ * @author Krzysztof Jodlowski
+ */
+public enum Symbol {
+  NAUGHT("O"), CROSS("X"), EMPTY(" ");
```
Do you need EMPTY when you store marked fields in map?

KrzysiekJodlowski replied:
I used it as a placeholder in a few places but you're right - removed.

hongshui commented on readme.adoc:
```diff
@@ -13,8 +13,18 @@ To run project you have to use:
-
+- project with automated input run _"auto"_ script
+- running script from command line by typing *_./standard.sh_* or *_./auto.sh_*
+- if you can't run a script first execute command like: *_chmod 755 standard.sh_* (or *_chmod 755 auto.sh_*)
```
Just execute command not "commad LIKE"

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/model/Symbol.java:
```diff
@@ -0,0 +1,26 @@
+   */
+  @Override
+  public String toString() {
+    return this.value;
```
Just a quick reminder of my personal crusade against excessive usage of "this" keyword 😄.

hongshui commented on readme.adoc:
```diff
@@ -13,8 +13,18 @@ To run project you have to use:
+- project with automated input run _"auto"_ script
+- running script from command line by typing *_./standard.sh_* or *_./auto.sh_*
+- if you can't run a script first execute command like: *_chmod 755 standard.sh_* (or *_chmod 755 auto.sh_*)
+- when running _"auto"_ script new folder with name matching DATE_TIME name in automata/output is created, inside that folder you can see effect of running the game with all provided default scenarios:
```
results or effects - there are many of them

Maybe instead of "effect of running th game with all provided default scenarios" something simpler like "automated tests' results"?

KrzysiekJodlowski replied:
That sounds definitely better!

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/ui/ConsoleUI.java:
```diff
@@ -0,0 +1,59 @@
+    } while (!correctInput);
+    return input;
+  }
+}
```
Just a quick reminder of my personal crusade against excessive usage of "this" keyword 😄.

KrzysiekJodlowski replied:
And I will fight for it till the last drop of my blood:P No, serious, I begin to consider this habit..

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/ui/UI.java:
```diff
@@ -0,0 +1,27 @@
+   * @return custom Number type
+   */
+  T2 getNumberFromUser(NumberRange<T2> range);
+}
```
In my opinion too few methods. You can only show a message and get number from input. No option for showing error for Console UI to error stream for example (it could be error method and console ui implementation could do the same as say() but with printing to System.err not System.out).

KrzysiekJodlowski replied:
Thats a neat idea! I'll add error showing to UI:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
-        Board board;
-        int boardSize = 3;
+  /**
+   * Creating Game and runs it. First it checks if there is any input file and
```
You are using creat-ing with run-s. You shoud use creat-es instead. Also, maybe word "initializes" would be more fitting here?

KrzysiekJodlowski replied:
Initializes sounds the best:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
+            new FileInputStream(new File(args[0]))
+        );
+      } catch (FileNotFoundException e) {
+        System.out.println("File not found!");
```
Instead of catching exception, could you just check if file exists before trying to read from it? I think there is method for this

KrzysiekJodlowski replied:
exists() method, but to use it I must create File object which constructor can throw IOException which I'd must catch. We're running in the circle;)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
+    // helper method checking UI chosen type will initialize the interface
+    UI<String, Integer> ui = new ConsoleUI(scanner, new PrintStream(System.out));
+    Game game = new OxConsoleGame(ui);
+    game.run();
```
Could be one line.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
+   * @param args file with custom input (expecting file content as first parameter)
+   */
+  public static void main(final String[] args) {
+    Scanner scanner = setScanner(args);
```
Set suggests you have some private fied that is returned here (at least by Tomek's standards), so maybe change "set" to something else?

KrzysiekJodlowski replied:
"init":)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Move.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class Move implements Event<Move> {
+  private final FieldNumber fieldNumber;
+  private final Symbol playersSymbol;
```
Do you need symbol? BoardValidator has reference to Board object so you can get symbol calling board.getSymbol(move.getFieldNumber()).

KrzysiekJodlowski replied:
Yes, but first proper Symbol must be putted into board:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Event.java:
```diff
@@ -0,0 +1,16 @@
+   *
+   * @return the stored data
+   */
+  T getData();
```
No usage of this method.

KrzysiekJodlowski replied:
Removed, interface described as marker interface.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/App.java:
```diff
@@ -1,47 +1,50 @@
-            System.out.println(board);
-        }
+  private static Scanner setScanner(String[] args) {
+    Scanner scanner = null;
```
Instead of initializing scanner with null, you could just return new Scanner(new File...) or new Scanner(System.in), though this is personal preference

KrzysiekJodlowski replied:
Yes! And avoid else statement. Corrected:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/Event.java:
```diff
@@ -0,0 +1,16 @@
+package com.github.krzysiekjodlowski.ox;
+
+/**
+ * Describes generic event, and it's associated meta data which get
```
No "," before "and"

"metadata" is one word

"," can be before "which"

"which" could be changed to "that"

"get" to "gets" or "is"

KrzysiekJodlowski replied:
That one has changed now: "Describes generic event, only a marker interface."

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/Event.java:
```diff
@@ -0,0 +1,16 @@
+
+/**
+ * Describes generic event, and it's associated meta data which get
+ * sent in the bus to be dispatched to interested Subscribers.
```
I think send "via" or "by" but not "in the", though I am not 100% sure

KrzysiekJodlowski replied:
Your'e right, aah these ready solutions and their english:P

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/Event.java:
```diff
@@ -0,0 +1,16 @@
+ */
+public interface Event<T> {
+  /**
+   * Returns the stored data associated with the event.
```
Without "the stored"

KrzysiekJodlowski replied:
That one is a past;)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/EventBus.java:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Krzysztof Jodlowski
+ */
+public interface EventBus {
```
I really like the idea of this EventBus :)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/HumanPlayer.java:
```diff
@@ -0,0 +1,44 @@
+import com.github.krzysiekjodlowski.ox.ui.UI;
+
+/**
+ * ${@inheritDoc} Human version.
```
Nice use of inheritDoc :)

KrzysiekJodlowski replied:
:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/HumanPlayer.java:
```diff
@@ -0,0 +1,44 @@
+   * @return human Player representation
+   */
+  @Override
+  public String toString() {
+    return String.format("%s (human)", super.toString());
+  }
```
I've already written this but I don't think that specifying if someone is human or not is necessary

KrzysiekJodlowski replied:
It will be important when randomized computer player will enter the game;)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/HumanPlayer.java:
```diff
@@ -0,0 +1,44 @@
+   *
+   * @param ui generic user interface
+   * @return move made by player
+   * @throws NumberLowerThanOneException when want to mark field lower than 1
```
"when want" does not sound very well, maybe "after attemting" or something?

KrzysiekJodlowski replied:
According to above comments human doesn't make moves now.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/HumanPlayer.java:
```diff
@@ -0,0 +1,44 @@
+    NumberRange<Integer> boardBoundaryFields = NumberRange.of(
+        Settings.FIRST_FIELD_NUMBER,
+        Settings.INSTANCE.boardCapacity()
+    );
```
Doeas it change with every move? Because based on names I would assume not (boardCapacity and firstField should be the same for every game). If I am right maybe consider moving this assigment to same static / final field or something?

KrzysiekJodlowski replied:
Gone:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/NumberRange.java:
```diff
@@ -1,56 +1,63 @@
+   *
+   * @param minimalValue alleged lower bound of range
+   * @param maximumValue alleged upper bound of range
+   * @return new NumberRange object
```
alleged?

KrzysiekJodlowski replied:
What I had in my mind?:P Removed

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/Player.java:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @author Krzysztof Jodlowski
+ */
+abstract class Player {
```
Maybe some Score field (you don't have it so - maybe some Score class too 😃 ).

KrzysiekJodlowski replied:
Yes! That is the plan for "the next" step of this project;) Score for the Player, Round for the Game, Result keeping track of the Scores, Settings downloading default settings and Languages...

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    this.ui = ui;
+  }
+
+  /**
+   * ATTENTION - Below implementation is temporary!
+   * Analogical one will appear in future Round class.
+   * Also all hardcoded Strings will be read from file by Settings.
+   */
```
I don't think information that implementation is temporary should be in javadoc comment. Normal comment (/* */) would be more that enough

KrzysiekJodlowski replied:
Good point!

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    this.setStartingPlayer();
+    this.oxGameEventBus.register(board);
+    this.oxGameEventBus.register(boardValidator);
+    this.oxGameEventBus.register(this.players);
```
Nice to see few event bus subscribers :)

KrzysiekJodlowski replied:
And there will be more!

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/board/Board.java:
```diff
@@ -0,0 +1,104 @@
+    return this.fields.size();
+  }
+
+  Symbol getSymbol(FieldNumber fieldNumber) {
+    return this.fields.get(fieldNumber);
+  }
```
Good but you can return Optional of Symbol and that will reduce IF statements in your code.
When you will do that you can also delete containsField method (if not contains then return Optional of null) but you use that method in checker so you need to change few things in code so it is not that necessary.

KrzysiekJodlowski replied:
Good idea! Generally I use getSymbol in checkers so I'll remember about that in next iteration.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    } else {
+      return Settings.MIN_GAME_BOARD_SIDE_LENGTH;
+    }
+  }
```
Both functions are pretty difficult to understand in their current shape. You could try to simplify them

KrzysiekJodlowski replied:
I would! They'll be reimplemented in Round class.

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/ui/ConsoleUI.java:
```diff
@@ -0,0 +1,59 @@
+ * @author Krzysztof Jodlowski
+ */
+public class ConsoleUI implements UI<String, Integer> {
+  private final Scanner scanner;
+  private final PrintStream printStream;
```
Error handling needs some love. Please add PrintStream for error showing (you can pass System.err from main) 😺. I hope you read my comment in UI class, there is some additional info.

KrzysiekJodlowski replied:
That's done! Basically what I've done is a little walkaround by using the same stream and only changing the color of the String to red.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    this.ui.say("Which player should start? (enter 1 if player 'O' and 2 if player'X')");
+    int playerNumber = this.ui.getNumberFromUser(NumberRange.of(1, 2));
+    if (playerNumber == 2) {
+      this.players = new Players.Builder().changeStarting().build();
```
Why do you have to call builder to do this and why you call it here?

KrzysiekJodlowski replied:
Yes, not a Builder should do this. I've moved this method to Players, at least temporarily.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  private void showBoardAndSayItsOver(Board board, BoardValidator boardValidator) {
+    this.ui.say(board.toString());
```
You are calling this.ui.say(boad.toString()) few times, maybe extract it to separate method (e.g. printBoard())?

KrzysiekJodlowski replied:
It will look better in Round:)

AleksanderStudnicki commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+    this.ui.say("*** Welcome to OX game! ***");
+  }
+
+  private Board setBoardSideLength() {
```
"Set" as method name prefix but it returns Board object not sets. Doomguy is not happy :rage1: .

KrzysiekJodlowski replied:
Thats changed to getBoardSideLengthFromUser, assigned to int and Board is initialized separately now:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxConsoleGame.java:
```diff
@@ -0,0 +1,110 @@
+      this.ui.say("It's a draw!");
+    } else {
+      this.ui.say(String.format("The winner is %s!", boardValidator.getWinner()));
+    }
```
This two could be separate methods. Also consider that when you will have more than one match draw / win will be said more times that "game over"

KrzysiekJodlowski replied:
As I said, when legendary Round class will appear these methods will be torn apart and rebiuld:)

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/OxGameEventBus.java:
```diff
@@ -0,0 +1,32 @@
+ */
+class OxGameEventBus implements EventBus {
+
+  private final List<Subscriber> subscribers = new ArrayList<>();
```
You could use LinkedList here

KrzysiekJodlowski replied:
Especially that I'm using forEach here - changed!

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/Player.java:
```diff
@@ -0,0 +1,61 @@
+ * @author Krzysztof Jodlowski
+ */
+abstract class Player {
+  private final Symbol playersSymbol;
```
symbol instead of playersSymbol would be enough. We are in Players class after all, everything here is "player's"

KrzysiekJodlowski replied:
Good point! Changed.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/Players.java:
```diff
@@ -0,0 +1,61 @@
+  private final Player playerTwo;
+  private Player currentPlayer;
+
+  private Players(Builder builder) {
+    this.playerOne = builder.playerOne;
+    this.playerTwo = builder.playerTwo;
+    this.currentPlayer = builder.startingPlayer;
+  }
```
You have only three parameters, all of them required. Do you need builder for this?

KrzysiekJodlowski replied:
Not for now, but there will be also option for setting the name of each player:)

AleksanderStudnicki reviewed (commented):
Game is working properly but project has some design flaws. I marked them as the comments.
Consider changes. Some of them are minor, some of them are little bit bigger but I'm sure you can do that.

Also few changes for pom.xml:
- add reporting section
- add reporting plugins
- update enforcer plugin

I pushed a new branch to the repository ("reviewAlek") with updated pom. You can just copy that to yours.

PS: README has an older version (0.1 instead of 0.2).

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/Board.java:
```diff
@@ -0,0 +1,104 @@
+
+  public Board(final int boardSideLength) {
+    this.boardSideLength = boardSideLength;
+    this.boardCapacity = this.boardSideLength * this.boardSideLength;
```
You have boardCapacity both in settings and here? That's confusing...

KrzysiekJodlowski replied:
Removed from settings:) boardSideLength also.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/BoardValidator.java:
```diff
@@ -0,0 +1,65 @@
+  private VerticalChecker verticalChecker;
+  private SlantChecker slantChecker;
+  private ReverseSlantChecker reverseSlantChecker;
+  private Board board;
```
You have a lot of fields in one class. It is usually not a good sign. So can check if there is possiblility to change that somehow (I can see four checkers here so I guess they have to stay but you can try to remove isGameOver, winner and board)

KrzysiekJodlowski replied:
isGameOver and winner will be kept in the future Result class, probably on next iteration;)

AleksanderStudnicki reviewed (changes requested):
Same as previous but changes requested.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/BoardValidator.java:
```diff
@@ -0,0 +1,65 @@
+
+  public boolean saysItsOver() {
+    return this.isGameOver;
+  }
```
Validator does not say anything. It just returns boolean so method could be named "isGameOver" and variable (if it has to exist) could be simple gameOver or somehting

KrzysiekJodlowski replied:
Changed!

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/HorizontalChecker.java:
```diff
@@ -0,0 +1,59 @@
+      }
+    }
+    return found >= this.winCondition;
+  }
```
This method could be 10 smaller

KrzysiekJodlowski replied:
And will be! Im planning to reimplement checkers with functional interfaces.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/HorizontalChecker.java:
```diff
@@ -0,0 +1,59 @@
+    return found >= this.winCondition;
+  }
+
+  private boolean checkIfThereIsAnother(int potentialCorner, int nextField, Move playersMove)
```
Another what?

KrzysiekJodlowski replied:
Symbol of course;) Changed!

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/board/ReverseSlantChecker.java:
```diff
@@ -0,0 +1,67 @@
+      }
+    }
+    return found >= this.winCondition;
+  }
```
Same, very long method

KrzysiekJodlowski replied:
That will change in next edition.

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/ui/ConsoleUI.java:
```diff
@@ -0,0 +1,59 @@
+      }
+    } while (!correctInput);
+    return input;
+  }
```
This method could be made into few smaller ones

KrzysiekJodlowski replied:
I've splitted it into two, the first part is a bit too problematic (returning String / catching Exception / loop flow control).

hongshui commented on src/main/java/com/github/krzysiekjodlowski/ox/ui/UI.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Interacting directly with user,
+ * used for collecting input end
+ * outputting messages to user.
```
outputting?

KrzysiekJodlowski replied:
Corrected:P

hongshui commented on src/test/java/com/github/krzysiekjodlowski/ox/BoardValidatorHorizontalCheckTest.java:
```diff
@@ -0,0 +1,219 @@
+
+        //assert
+        assertTrue(isGameOver);
+    }
```
This tests are very difficult to understand. 

Also, they seem very similar so you could try to extract some common parts.

KrzysiekJodlowski replied:
They will be splitted into classes with DataProviders and Board / Move / FieldNumbers will be mocked as it should be.

hongshui commented on src/test/java/com/github/krzysiekjodlowski/ox/BoardValidatorHorizontalCheckTest.java:
```diff
@@ -0,0 +1,219 @@
+        Board board = new Board(40);
+        BoardValidator boardValidator = new BoardValidator(board, 40);
+
+        Arrays.stream(chosenFields).forEach(x ->
+                {
+                    try {
+                        board.handle(new Move(FieldNumber.valueOf(x), Symbol.CROSS));
+                        boardValidator.handle(new Move(FieldNumber.valueOf(x), Symbol.CROSS));
+                    } catch (NumberLowerThanOneException e) {
+                        e.printStackTrace();
+                    }
+                }
+        );
```
?

KrzysiekJodlowski replied:
I had no other way, you can't add throws claue to method signature when using streams:/

hongshui commented on src/test/java/com/github/krzysiekjodlowski/ox/ConsoleUITest.java:
```diff
@@ -20,21 +21,20 @@
+//                , {"1", NumberRange.of(2, 2)}
+//                , {"1", NumberRange.of(2, 3)}
+//        };
+//    }
```
Remember to uncomment or delete. You can create issue for this.

KrzysiekJodlowski replied:
Deleted for now.

hongshui commented on src/test/java/com/github/krzysiekjodlowski/ox/ConsoleUITest.java:
```diff
@@ -46,19 +46,19 @@ public void testGettingNumberFromUserWhileNumberInRange(
+//        consoleUI.getNumberFromUser(numberRange);
+//
+//        //assert throws
+//    }
```
Remember to uncomment or delete. You can create issue for this.

KrzysiekJodlowski replied:
Deleted for now.

hongshui commented on src/test/java/com/github/krzysiekjodlowski/ox/HumanPlayerTest.java:
```diff
@@ -0,0 +1,59 @@
+  
+  }               
+
+
```
Empty spaces

KrzysiekJodlowski replied:
Aah, deleted:)

hongshui reviewed (approved):
Remarks:
- Sometimes you use verbs ending on "-ing" another time you use "-s". I saw them mixed even in one sentence. Try to be consistent with whatever time you are using when writting you javadoc (I do recommend present time).
- In your checker classes you have a lot of code duplication. Try to split methods into smaller parts, then check what differs and what is the same and try do to something about it
- Your board tests are pretty difficult to understand, maybe try some different aproach in the future?
- I really like your idea with EventBus

KrzysiekJodlowski committed:
Refactor according to PR #4 comments.

KrzysiekJodlowski commented:
Thank you for kind in-depth review! If there is something more to change let me know:)

Commit:
Update python script

Add creating output folder.

Commit:
Update python script

And templates according to PR updates.

Commit:
Update .gitignore

AleksanderStudnicki reviewed (approved):
Good.

PS. Unused methods in Players.Builder

Commit:
Final refactor according to PR #4

- add checkstyle plugin to build (to provide google_checks on plugin
call)
- remove unused methods in Players
- add static keyword to Settings according to checkstyle
